docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+```go
+type ICipher interface {
+  Encrypt(message Blob) (encryptedMessage Blob, error)
+  Decrypt(encryptedMessage Blob) (message Blob, error)
+}
+```
+
+#### Hasher
+
+This package contains the different hashing algorithms and conventions agreed on this matter.
+
+```go
+type IHasher interface {
+  Hash(input Blob) Blob
+  CanHashIncrementally() bool
+}
+```
+
+### Secure Storage
+
+A *Secure Storage* represents a secure vault where one or more *Secure Items* can be stored. As a parallelism, a `SecureStorage` instance can be compared to the current `Keyring Backend` interface,
+but with the main difference that the stored items are abstracted as `SecureItem` instead of `Record`. This allows a fully decoupled relation between what is being stored (`SecureItem`) and where it is being stored (`SecureStorage`).
+To access a *Secure Item*, users must interact with the *Secure Storage*, which handles the retrieval and management of keys.
+Different implementations of *Secure Storage* will be available to cater to various storage requirements:
+
+* FileSystem: This implementation stores the items in a designated folder within the file system.
+* Memory: This implementation stores the items in memory, providing fast access but limited persistence (for testing purpose mainly).
+* KMS: This implementation utilizes the Key Management System available on AWS, GCP, etc.
+* others: 1password, OS-integrated secure storage (macOS, Linux, Windows)
+
+```go
+type ISecureStorage interface {
+  List() []string
+
+  Get(name string) (SecureItem, error)
+  Set(item SecureItem) error
+  Remove(name string) error
+}
+```
+
+### Secure Item
+
+A *Secure Item* is a structured data object designed for storing any type of data within a *Secure Storage* instance.
+In the context of this ADR, the `Blob` field of a `SecureItem` represents the serialized form of a `CryptoProvider`.
+
+_Note:_ The encoder/decoder of the `Blob` field will be dependent on the `CryptoProvider` implementation, a `SecureItem`
+has no knowledge of the underlying data structure.
+
+```go
+type ISecureItemMetadata interface {
+  Type()   TypeUUID     // Relates to the corresponding provider
+  Name()   string
+}
+
+type ISecureItem interface {
+  ISecureItemMetadata
+  Bytes() []byte
+}
+```
+
+##### Keyring
+
+The `Keyring` interface serves as a central hub for managing `CryptoProviders` and `SecureStorage` implementations.
+The main difference with the current `Keyring` interface is that it adds the ability to register several `SecureStorage` (this is, several "storing backends") implementations and the
+_values_ are abstracted as `SecureItem` instead of `Record`.
+
+```go
+type IKeyring interface {
+  RegisterCryptoProviderBuilder(typeUUID TypeUUID, builder CryptoProviderBuilder)
+  RegisterAndLoadStorageProvider(typeUUID TypeUUID, provider StorageProvider)
+
+  ListStorageProviders() ([]IStorageProvider, error)
```

Comment:
`IStorageProvider` is not defined anywhere. Is it an old name for `SecureStorage`?